### ISSUE TITLE
Memory leak when iterating through las files

### ISSUE THREAD

**Issue by jeffreywolf** [grantbrown/laspy#32]
This is a great library. Thanks to the author for your hard work putting this together.

I've been toying around with it for applications in Ecology. One of the problems I've been working on requires either sequential or parallel processing of hundreds of las files (I rasterize the point clouds into a digital elevation model and digital surface model).

I have run into an issue with iterating over a list of files [i.e., `glob.glob('*.las')`]. There are some code blocks below, but what follows is a general description of what happens. 

I run the top command and watch system resources in a separate terminal from python. Memory allocated to Python continually increases as I sequentially iterate over the filenames, open the files in read mode, use data from the files, close the file, and repeat.  I first noticed the problem in a script, but the same thing happens when running commands from within ipython. When I use dir() to list the objects in the main namespace in Python after already closing the las file there are still references to the data objects I used from the file.  Trying to use these data objects at this point results in a segmentation fault. 

I'm not sure what the best way is to make a reproducible result, because that would require transferring the las files somewhere. I expect the same thing will happen with any collection of las files though.

I am using laspy 1.2.5 / 64bit Anaconda on Mac OS X. Some toy code follows,

```
import glob, numpy as np, laspy

#assumes you are in a directory with a collection of las files
files = glob.glob("*.las")
for i, f in enumerate(files):
    #print the number of the file in the iteration and the filename
    print i, f
    # open file
    las = laspy.file.File(f, mode="r")
    #use point data
    points = las.points
    #do something with point data -- trivial example here is printing points
    print points
    #close las file
    las.close()
```

Something similar, but perhaps more insightful about what is going on follows. Warning... this crashes on my system.

```
import glob, numpy as np, laspy

#assumes you are in a directory with a collection of las files
files = glob.glob("*.las")

f0 = files[0]
las = las.file.File(f0, mode="r")
points = las.points
las.close()
#list objects in main namespace
dir()
#points is still present in the namespace
#calling points now will crash python with a segfault
points
```

**Comment by hobu** [comment]
I assume you're just running out of memory, and there's some kind of reference counting-related bug in laspy that's causing stuff to hang on and not be deleted.

**Comment by grantbrown** [comment]
Points are provided using numpy frombuffer, so closing the file (and therefore the underlying map) and then trying to access the points is going to cause issues. If you're certain that you want to keep point data around after closing a file, you could do something like:

    import glob, numpy as np, laspy

    #assumes you are in a directory with a collection of las files
    files = glob.glob("*.las")

    f0 = files[0]
    las = las.file.File(f0, mode="r")
    points = las.points.copy()
    las.close()
    #list objects in main namespace
    dir()
    #points is still present in the namespace
    #calling points now should not result in a crash
    points

An even better solution would be to use a function to explicitly define what data you want out of each file before closing them.

    import glob, numpy as np, laspy

    #assumes you are in a directory with a collection of las files
    files = glob.glob("*.las")

    def process(fname):
        las = las.file.File(fname, mode="r")
        points = las.points.copy()
        las.close()
        return(points)

    for i, f in enumerate(files):
        print i, f
        pts = process(f)
        print points


I'd be interested to know if you continue to have problems with an approach like that.

**Comment by jeffreywolf** [comment]
Thanks (@hobu/@grantbrown) for the quick reply.

I tried the last code block (made a few minor changes: module name in process changed to 'laspy',  'points' in for loop changed to pts")...

```
import glob, numpy as np, laspy

#assumes you are in a directory with a collection of las files
files = glob.glob("*.las")

def process(fname):
    las = laspy.file.File(fname, mode="r")
    points = las.points.copy()
    las.close()
    return(points)

for i, f in enumerate(files):
    print i, f
    pts = process(f)
    print pts
```

The problem persists. Memory is still not deallocated. Here is a screenshot after passing through 141 files (a subset of all the files I am working with). I ran this on a 64GB RAM server. There are >250 las files in the directory.

![laspy](https://cloud.githubusercontent.com/assets/3539624/9937119/4c2c6c22-5d2d-11e5-97a0-35565aa4b397.png)

**Comment by jeffreywolf** [comment]
I don't know whether the problem is with laspy.  Has anyone iterated over many las files using laspy for i/o and monitored memory usage before?

**Comment by grantbrown** [comment]
I'm looking into the issue.

**Comment by jeffreywolf** [comment]
Thank you! Glad to help in any way, but I'm not familiar with the code-base.

**Comment by grantbrown** [comment]
Following the instructions [here](https://pythonhosted.org/Pympler/muppy.html), I've managed to reduce the size of the memory leak in [this](https://github.com/grantbrown/laspy/commit/eed5b3d6965dbc39878370ea213abba73267ee91) commit on a feature branch. There are still quite a few remaining objects, but I wasn't immediately able to figure out why or where they came from.

I'll have to return to this later, but if either of you run across the culprit in the meantime definitely let me know.

**Comment by grantbrown** [comment]
Leak check script:
    from pympler import tracker                                                     
    import laspy                                                                    
                                                                                                                                                      
    def g():                                                                        
        f = laspy.file.File("simple.las")                                           
        pts = f.points.copy()                                                       
        del(pts)                                                                    
        f.close()                                                                   
        del(f)                                                                      
                                                                                
    memory_tracker = tracker.SummaryTracker()                                       
    memory_tracker.print_diff()                                                     
    print("########## START   ##########")                                           
                                                                                
    for i in xrange(1000):                                                          
        g()                                                                         
                                                                                                                                                   
    memory_tracker.print_diff()

**Comment by KAMI911** [comment]
Hi,
I tested the the memory_tracker code, and I got OOM while testing on a 259653143 bytes long las file. I had to set xrange to 2 because 10 caused long-long wait and huge memory usage. 100..1000 caused OOM.

The output of the script:
```
python2.7 ./mem.py 
                  types |   # objects |   total size
======================= | =========== | ============
                   list |        2623 |    266.38 KB
                    str |        2625 |    150.42 KB
                    int |         272 |      6.38 KB
                   dict |           2 |      1.30 KB
     wrapper_descriptor |           7 |    560     B
                weakref |           3 |    264     B
      member_descriptor |           2 |    144     B
                   code |           1 |    128     B
  function (store_info) |           1 |    120     B
                   cell |           2 |    112     B
      getset_descriptor |           1 |     72     B
      method_descriptor |           1 |     72     B
                  tuple |           0 |      8     B
         instancemethod |          -1 |    -80     B
########## START   ##########
                               types |   # objects |   total size
==================================== | =========== | ============
                                 int |    18546627 |    424.50 MB
                                list |          23 |    155.51 MB
                                dict |          38 |     47.89 KB
                            instance |         188 |     13.22 KB
                                 str |         121 |      6.42 KB
                              Struct |          10 |      5.75 KB
                lxml.etree._Document |          10 |    880     B
                 lxml.etree._Element |          10 |    720     B
                               float |          24 |    576     B
                                 set |           2 |    464     B
                                file |           2 |    288     B
                         numpy.dtype |           2 |    176     B
  <class 'laspy.header.HeaderManager |           2 |    128     B
         <class 'laspy.header.Header |           2 |    128     B
```
I am using multiprocessing module to handle my transforming task here:
https://github.com/KAMI911/lactransformer/blob/master/lactransformer.py
Using https://github.com/KAMI911/lactransformer/blob/master/lib/TransformerWorkflow.py and https://github.com/KAMI911/lactransformer/blob/master/lib/LasPyConverter.py modules. Probably using multiprocessing for laspy process can avoid the leak hence can be a workaround until the developers found the problem. Also I will try memory_tracker with my code. :+1:

**Comment by grantbrown** [comment]
Thanks for the help - I'll try to find some time this weekend to dig back into this issue.

**Comment by grantbrown** [comment]
Just to keep folks posted, I haven't forgotten about this issue, but am pretty strapped for time right now. I'll try to get back to this by next weekend. In the meantime, I'd be curious to know if anyone learns anything about the cause of the problem.

**Comment by jeffreywolf** [comment]
To keep folks posted from my end (e.g., laspy user) my workaround was to write a driver program that uses multiprocessing and the subprocess library to run each las tile in parallel through subprocess (command line) calls to an outside program.  The outside program called by subprocess opens (using laspy) a single 1 km x 1 km las tile  and rasterizes that point cloud (using numpy and GDAL).  The memory is freed when each subprocess finishes, so I was able to process a collection of over >250 1 km x 1 km las tiles this way.   This may be helpful information for others interested in using laspy for processing multiple las files while the issue is being resolved.

**Comment by WilliamWalker** [comment]
Yes, great library.  I too have recently encountered the following error after I open/close several LAS files consecutively:  Error mapping file: [Error 8] Not enough storage is available to process this command

While monitoring my system resources, it appears that memory is allocated when a file is opened (e.g. inFile = laspy.file.File(las_path, mode='r')), but is not released with inFile.close().  If i restart my python shell, the memory is released.  Any update on this would greatly appreciated.

**Comment by grantbrown** [comment]
This issue definitely needs to be fixed, however I haven't yet had a chance to revisit it. In the meantime, using the multiprocessing trick mentioned by KAMI911 and jeffrywolf should hopefully help (as sub-optimal as this is).

From my initial research, the leak may arise from circular references between several of the worker classes in laspy (Header, HeaderManager, Reader etc.)

**Comment by grantbrown** [comment]
I think I've contained the worst of the problem with [bb41e5](https://github.com/grantbrown/laspy/commit/bb41e5b57fd168c4b71427def2b0fc78af1803eb). It looks like there may still be a few bytes around after file close, but it doesn't seem to scale with the number of files processed. I'd be grateful if those of you who've had issues could test out the latest changes to see if you still run into problems (there could certainly be other problematic code paths). 

Sorry it's taken me so long to return to the issue.

**Comment by jonathanvoss** [comment]
Does that actually work? You know __del__ is only ever called when the garbage collector is reclaiming that object, right? As far as I can tell, deleting references in __del__ should be no different from the normal behavior of a Python object.

**Comment by grantbrown** [comment]
You're right, I didn't continue testing to find the minimal subset of changes sufficient to deal with the leak. Looking again, it turns out that the problem was that I'd defined a __del__ method already, which simply closed the file without doing the rest of the work the default object behavior. Refactored [here](https://github.com/grantbrown/laspy/commit/8d8bc7a8c139dae09f39e266df72619f8a833eff), thanks.

**Comment by jonathanvoss** [comment]
Ah, I deleted those in my branch, as context managers are a better way to make sure files get closed at the right time.

**Comment by grantbrown** [comment]
Marking issue resolved, version bumped to 1.4.1, pypi updated
